### ISSUE TITLE
How to constrain the zooms factors to exactly the provided tile imagery?

### ISSUE THREAD

**Issue by rburkey** [openseadragon/openseadragon#556]
I have a custom tile source consisting of 8 zoom levels (1:1, 2:1, ..., 128:1), levels 7 through 0, and I'd like to constrain the zoom levels that are displayed in the viewport to exactly those resolutions.  In other words, I'd like everything to be displayed at exactly 1 displayed pixel per each tile pixel, rather than having the viewer do any scaling.  I can't seem to figure out any way to do to that.  Is there any way?

The best I'm able to do is the following:
  maxZoomPixelRatio: 1.0,
  zoomPerScroll: 2.0,
  minLevel: 0,
  maxLevel: 7

With these settings, if I zoom all the way in, then zoom outward one step at a time, the first 8 views do correspond to the 8 resolutions in my tiles ... but then the viewer lets me zoom out three more times, to resolutions having nothing to do with my tiles.  If I then zoom inward step by step, none of the resolutions correspond to my tiles until I'm zoomed in fully again.

**Comment by msalsbery** [comment]
maxZoomLevel would be another setting, but I don't think it currently does anything.

Don't know if this is an option, but this case is a big reason I made the [OpenSeadragonViewerInputHook plugin](https://github.com/msalsbery/OpenSeadragonViewerInputHook). I hook zoom related gestures on the viewer and do the calculating/zooming myself.

**Comment by rburkey** [comment]
I looked at the OpenSeadragonViewerInputHook (actually, just a few minutes ago) to see if it could help me, but didn't understand how it might.  I'll look again, thanks!

**Comment by rburkey** [comment]
One gotcha that hit me immediately is in $.ViewerInputHook.version in viewerinputhook.js.  I guess (not knowing anything about git) that 

        versionStr: '<%= viewerinputhookVersion.versionStr %>',
        major: <%= viewerinputhookVersion.major %>,
        minor: <%= viewerinputhookVersion.minor %>,
        revision: <%= viewerinputhookVersion.revision %>

are supposed to be automatically replaced by git.  But if you download the zipfile (which is what I did), they don't get replaced, and so the resulting file has a syntax error and doesn't work.  I replaced it with 

        versionStr: 'Version 1.1.0',
        major: 1,
        minor: 1,
        revision: 0

which at least gets rid of the syntax error.

**Comment by msalsbery** [comment]
If you download a zip on GitHub then you'd have to build it using grunt. I should include builds with the source :) Sorry about that.

The readme has links to minified and dev builds.=

**Comment by rburkey** [comment]
Ah!  Well, I admit that when you're trying to get in and get out without learning anything about it --- as I was --- you tend to overlook details like that.  :-)

At any rate, it appears to work, even though it was a bit of a chore to work through all the details for this specific use-case.  Thanks!

**Comment by msalsbery** [comment]
Yeah, I neglected to mention that it can be a bit of work. The source code for the default gesture handlers in the OpenSeadragon.Viewer class are a good starting point and they are relatively small. That just leaves the custom math, which can be a bit confusing.

I've worked a lot with microscopy images that have levels which are actually scanned at different magnification so I need similar functionality - clear steps between zoom levels. With a viewer that's resizable the math gets a little trickier since OpenSeadragon's zoom levels are a factor of the viewer size. That's what spawned that set of plugins :)

**Comment by iangilman** [comment]
Should we add a "stop at tile levels" zoom setting for OSD? Or have a drop-in plugin that does this?

**Comment by rburkey** [comment]
Well, if you're asking for the opinion of a casual user who doesn't expect to delve into the guts of the library, there are a couple of viewer settings I expected to find and didn't.  The setting you mention is one of them.

But ... what would really have helped me out in my particular use-case would have been if the viewer element simply had an setting that was a list of allowed zoom-levels and always stuck to those specific levels *if* the setting was present, and had its current behavior otherwise.   A setting like that would have turned my implementation effort into an hour rather than a couple of days.  (Not that there was very much coding involved, but all of the googling, the messaging, the experimenting with settings that didn't work quite like the documentation suggested, etc., turned it into more of an adventure than I hoped for.  On the other hand, the end result was quite nice!)

**Comment by rburkey** [comment]
Incidentally, before I forget, I found a bug (at least in my opinion) in the OpenSeadragonImagingHelper plugin.  The zoomOut and zoomOutAboutLogicalPoint methods are not inverses of the corresponding zoomIn methods, and therefore produce different zoom factors moving outward than are produced moving inward.  Here's the change I had to make in zoomOutAboutLogicalPoint, which was the only one of those methods I happened to be using:

            //newzoom *= (1.0 - this._zoomStepPercent / 100.0);
            newzoom /= (1.0 + this._zoomStepPercent / 100.0);

For a tiny zoomStepPercent like 1 or 10, the results would be close, but for 100 (which I was obliged to use), the results were completely bogus.

**Comment by msalsbery** [comment]
>I found a bug (at least in my opinion) in the OpenSeadragonImagingHelper plugin

@rburkey Thanks for pointing that out! I'll patch that right away!

**Comment by msalsbery** [comment]
@rburkey If you're interested in contribution credit on GitHub you're welcome to file a bug on [the plugin's issues page](https://github.com/msalsbery/OpenSeadragonImagingHelper/issues). If not, I'll take care of it. Thanks again!

**Comment by rburkey** [comment]
I'm too lazy for that, thanks.  :-)

________________________________________

**Comment by msalsbery** [comment]
Hehe no problem :)

**Comment by iangilman** [comment]
@rburkey Thanks for the feedback! Hopefully someone will take this on.

**Comment by msalsbery** [comment]
@rburkey  Fixed OpenSeadraagonImagingHelper plugin. 

Get version 1.2.0 here:
[openseadragon-imaginghelper.min.js](http://msalsbery.github.io/builds/openseadragonimaging/openseadragon-imaginghelper.min.js)
[openseadragon-imaginghelper.js](http://msalsbery.github.io/builds/openseadragonimaging/openseadragon-imaginghelper.js)
